picard/ui/matchdetailsdialog/__init__.py

```diff
@@ -0,0 +1,395 @@
+                if not self.isRowHidden(i):
+                    item = self.item(i, self.col)
+                    if item.text() not in data_unique:
+                        data_unique.append(item.text())
+
+            data_unique.sort()
+
+            for item in data_unique:
+                checkBox = QCheckBox(item, self.menu)
+                checkBox.setChecked(True)
+                checkableAction = QWidgetAction(self.menu)
+                checkableAction.setDefaultWidget(checkBox)
+                self.menu.addAction(checkableAction)
+                self.filter_checkboxes.append(checkBox)
+
+            btn = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel, QtCore.Qt.Horizontal, self.menu)
+            btn.accepted.connect(self.menuClose)
+            btn.rejected.connect(self.menu.close)
+            checkableAction = QWidgetAction(self.menu)
+            checkableAction.setDefaultWidget(btn)
+            self.menu.addAction(checkableAction)
+
+            headerPos = self.mapToGlobal(self.horizontalHeader().pos())
+
+            posY = headerPos.y() + self.horizontalHeader().height()
+            posX = headerPos.x() + self.horizontalHeader().sectionPosition(self.col)
+            self.menu.exec_(QtCore.QPoint(posX, posY))
+        except Exception:
+            log.error(traceback.format_exc())
+
+    def menuClose(self):
+        self.keywords[self.col] = []
+        did_filter = False
+        for element in self.filter_checkboxes:
+            if element.isChecked():
+                self.keywords[self.col].append(element.text())
+            else:
+                did_filter = True
+
+        # add decorator to show filtered.
+        if did_filter:
+            it = self.horizontalHeaderItem(self.col)
+            oldHeader = it.text()
+            newHeader = oldHeader + "*"
+            it.setText(newHeader)
+
+        self.filterdata()
+        self.menu.close()
+
+    def clearFilter(self):
+        # clear filter
+        for i in range(self.rowCount()):
+            self.setRowHidden(i, False)
+
+        # remove filter decoroator
+        for i in range(self.columnCount()):
+            it = self.horizontalHeaderItem(i)
+            oldHeader = it.text()
+            newHeader = re.sub('[*+]', '', oldHeader)
+            it.setText(newHeader)
+
+    def filterdata(self):
+
+        columnsShow = dict([(i, True) for i in range(self.rowCount())])
+
+        try:
+            for i in range(self.rowCount()):
+                for j in range(self.columnCount()):
+                    item = self.item(i, j)
+                    if j in self.keywords:
+                        if item.text() not in self.keywords[j]:
+                            columnsShow[i] = False
+            for key, value in columnsShow.items():
+                self.setRowHidden(key, not value)
+
+        except Exception:
+            log.error(traceback.format_exc())
+
+
+class SortableTableWidgetItem(QtWidgets.QTableWidgetItem):
```

Comment:
Duplicate of https://github.com/metabrainz/picard/blob/170b8e0aa70ee4a4c49e89ab42e26404e62e90dc/picard/ui/searchdialog/__init__.py#L172-L179

Reply:
Ideally all this logic would have already been in a super class that SearchDialog inherited from, then I could have used it in MatchDetailsDialog.  I just don't see messing with SearchDialog, to be good to do as part of this PR.  Let me know otherwise.

Reply:
> Ideally all this logic would have already been in a super class that SearchDialog inherited from

Well, until this PR, there was no reason to do so. Though we usually try to reduce code redundancy, and when such situation arises, we extract common stuff and make a preliminary PR introducing classes to be re-used in subsequent PR.

I did a quick comparaison, there are a lot of lines in common. You even introduced small changes that artificially decrease code duplication, like in `save_table_header_state()`, `restore_table_header_state()`, where the only change is the way you format debug message, one using sprintf-style the other format(), but with exact same message and parameter.

This is definitively something we want to avoid. So let's just make a preliminary PR, that moves common code you need in your new classes to separate classes you'll inherit from.

@phw what do you think?